collect_app/src/main/java/org/odk/collect/android/instancemanagement/InstanceAutoSender.kt

```diff
@@ -0,0 +1,97 @@
+package org.odk.collect.android.instancemanagement
+
+import android.content.Context
+import org.odk.collect.analytics.Analytics
+import org.odk.collect.android.R
+import org.odk.collect.android.formmanagement.InstancesAppState
+import org.odk.collect.android.gdrive.GoogleAccountsManager
+import org.odk.collect.android.gdrive.GoogleApiProvider
+import org.odk.collect.android.notifications.Notifier
+import org.odk.collect.android.upload.FormUploadException
+import org.odk.collect.android.utilities.ChangeLockProvider
+import org.odk.collect.android.utilities.FormsRepositoryProvider
+import org.odk.collect.android.utilities.InstanceUploaderUtils
+import org.odk.collect.android.utilities.InstancesRepositoryProvider
+import org.odk.collect.forms.FormsRepository
+import org.odk.collect.forms.instances.Instance
+import org.odk.collect.forms.instances.InstancesRepository
+import org.odk.collect.permissions.PermissionsProvider
+import org.odk.collect.settings.SettingsProvider
+import org.odk.collect.settings.keys.ProjectKeys
+import org.odk.collect.shared.settings.Settings
+
+class InstanceAutoSender(
+    private val context: Context,
+    private val changeLockProvider: ChangeLockProvider,
+    private val notifier: Notifier,
+    private val analytics: Analytics,
+    private val formsRepositoryProvider: FormsRepositoryProvider,
+    private val instancesRepositoryProvider: InstancesRepositoryProvider,
+    private val googleAccountsManager: GoogleAccountsManager,
+    private val googleApiProvider: GoogleApiProvider,
+    private val permissionsProvider: PermissionsProvider,
+    private val settingsProvider: SettingsProvider,
+    private val instancesAppState: InstancesAppState
+) {
+    fun autoSendInstances(projectId: String): Boolean {
+        val formsRepository = formsRepositoryProvider.get(projectId)
+        val instancesRepository = instancesRepositoryProvider.get(projectId)
+        val generalSettings = settingsProvider.getUnprotectedSettings(projectId)
+        val instanceSubmitter = InstanceSubmitter(
+            analytics,
+            formsRepository,
+            googleAccountsManager,
+            googleApiProvider,
+            permissionsProvider,
+            generalSettings
+        )
+        return changeLockProvider.getInstanceLock(projectId).withLock { acquiredLock: Boolean ->
+            if (acquiredLock) {
+                val toUpload = getInstancesToAutoSend(
+                    formsRepository,
+                    instancesRepository,
+                    generalSettings
+                )
+                try {
+                    val result: Map<Instance, FormUploadException?> = instanceSubmitter.submitInstances(toUpload)
+                    notifier.onSubmission(result, projectId)
+                } catch (e: SubmitException) {
+                    if (e.type == SubmitException.Type.GOOGLE_ACCOUNT_NOT_SET) {
```

Comment:
It looks like this doesn't handle the `NOTHING_TO_SUBMIT`? I'm guessing we don't do anything in that case (no notification) but either way I think making this a when on `SubmitException#type` and then explicitly "no-oping" in the `NOTHING_TO_SUBMIT` case would be clearer.

Reply:
Fixed.